### PR TITLE
[SPARK-13902][SCHEDULER] Make DAGScheduler not to create duplicate stage.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

`DAGScheduler`sometimes generate incorrect stage graph.

Suppose you have the following DAG:

```
[A] <--(s_A)-- [B] <--(s_B)-- [C] <--(s_C)-- [D]
            \                /
              <-------------
```

Note: [] means an RDD, () means a shuffle dependency.

Here, RDD `B` has a shuffle dependency on RDD `A`, and RDD `C` has shuffle dependency on both `B` and `A`. The shuffle dependency IDs are numbers in the `DAGScheduler`, but to make the example easier to understand, let's call the shuffled data from `A` shuffle dependency ID `s_A` and the shuffled data from `B` shuffle dependency ID `s_B`.
The `getAncestorShuffleDependencies` method in `DAGScheduler` (incorrectly) does not check for duplicates when it's adding ShuffleDependencies to the parents data structure, so for this DAG, when `getAncestorShuffleDependencies` gets called on `C` (previous of the final RDD), `getAncestorShuffleDependencies` will return `s_A`, `s_B`, `s_A` (`s_A` gets added twice: once when the method "visit"s RDD `C`, and once when the method "visit"s RDD `B`). This is problematic because this line of code: https://github.com/apache/spark/blob/8ef3399/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L289 then generates a new shuffle stage for each dependency returned by `getAncestorShuffleDependencies`, resulting in duplicate map stages that compute the map output from RDD `A`.

As a result, `DAGScheduler` generates the following stages and their parents for each shuffle:

|  | stage | parents |
| --- | --- | --- |
| s_A | ShuffleMapStage 2 | List() |
| s_B | ShuffleMapStage 1 | List(ShuffleMapStage 0) |
| s_C | ShuffleMapStage 3 | List(ShuffleMapStage 1, ShuffleMapStage 2) |
| - | ResultStage 4 | List(ShuffleMapStage 3) |

The stage for s_A should be `ShuffleMapStage 0`, but the stage for `s_A` is generated twice as `ShuffleMapStage 2` and `ShuffleMapStage 0` is overwritten by `ShuffleMapStage 2`, and the stage `ShuffleMap Stage1` keeps referring the old stage `ShuffleMapStage 0`.

This patch is fixing it.
## How was this patch tested?

I added the sample RDD graph to show the illegal stage graph to `DAGSchedulerSuite`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -403,32 +403,47 @@ class DAGScheduler(
     parents.toList
   }
 
-  /** Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet */
-  private def getAncestorShuffleDependencies(rdd: RDD[_]): Stack[ShuffleDependency[_, _, _]] = {
-    val parents = new Stack[ShuffleDependency[_, _, _]]
+  /**
+   * Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet.
+   * This is done in topological order to create ancestor stages first to ensure that the result
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Can you move this comment to some place inside the method? It doesn't seem relevant to someone using this method, since the order of the ShuffleDependencies returned doesn't matter

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -322,6 +322,59 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with Timeou
...
+   *                               <--------------------
+   *                             /                       \
+   * [A] <--(1)-- [B] <--(2)-- [C] <--(3)-- [D] <--(4)-- [E] <--(5)-- [F]
+   *                \                       /
+   *                  <--------------------
+   *
+   * then check if all stages have correct parent stages.
+   * Note: [] means an RDD, () means a shuffle dependency.
+   */
+  test("parent stages") {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Can you add the JIRA name to this test?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 3]
```diff
@@ -322,6 +322,59 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with Timeou
     assert(sparkListener.stageByOrderOfExecution(0) < sparkListener.stageByOrderOfExecution(1))
   }
 
+  /**
+   * This test ensures that DAGScheduler build stage graph correctly.
+   * Here, we submit an RDD[F] having a linage of RDDs as follows:
+   *
+   *                               <--------------------
+   *                             /                       \
+   * [A] <--(1)-- [B] <--(2)-- [C] <--(3)-- [D] <--(4)-- [E] <--(5)-- [F]
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Can you update the JIRA to have a description of this test (this ascii art showing the DAG is great), and to explain what the previous (and incorrect) behavior was for this case?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -403,32 +403,47 @@ class DAGScheduler(
...
-            case _ =>
+      if (visited(r)) {
+        waitingForVisit.pop()
+      } else {
+        val visitedShuffleDeps = new ArrayBuffer[ShuffleDependency[_, _, _]]
+        val unvisitedDeps = new ArrayBuffer[Dependency[_]]
+
+        r.dependencies.foreach {
+          case dep: ShuffleDependency[_, _, _] if !shuffleToMapStage.contains(dep.shuffleId) =>
+            if (visited(dep.rdd)) visitedShuffleDeps += dep
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: The spark style (https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide) is to always put if/else statements in curly-braces, even if they're only one line

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 5]
```diff
@@ -403,32 +403,47 @@ class DAGScheduler(
...
+            else unvisitedDeps += dep
+          case dep if !visited(dep.rdd) => unvisitedDeps += dep
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Can you put the "unvisitedDeps += dep" on the next line (and indented)? AFAIK that's more consistent with the style used elsewhere in Spark

**Comment by markhamstra** [comment]
I think some of the terminology used in this and related PRs is confusing the issues.  When @kayousterhout and I ask about "correctness", what we are fundamentally concerned about is whether evaluation of the DAG produces the correct data elements.  I don't think that your description of "incorrect" or "illegal" graphs is meant to imply that incorrect data is produced from their evaluation.  Correct me if I am wrong, but I think that you are talking exclusively about graphs that are not optimal, causing duplication of effort and preventing further optimizations -- graphs that are taking longer to evaluate than is necessary, not graphs that are producing incorrect data elements.

If I am thinking correctly about this, then the entire effect of this and related PRs is to improve or optimize the DAGScheduler, not to create graphs and schedules that produce different end results than the DAGScheduler does now.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 3 above)
Comment (ueshin) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I updated the JIRA description.

**Comment by ueshin** [comment]
@markhamstra Thank you for your comment.
I thought the non-optimal state in `DAGScheduler` was a kind of bug so I used the words "incorrect" or "illegal" but now I see what you thought.
So yes, I was talking about only graphs and want to improve `DAGSchduler` performance by this and #12060.

**Comment by kayousterhout** [comment]
I just updated the [JIRA](https://issues.apache.org/jira/browse/SPARK-13902) with what I understand to be the issue.  Can you take a look and let me know if that's correct? If the simpler example I showed is sufficient to reproduce the issue (and my explanation is correct), can you simplify the unit test to use that example, and also update the JIRA and pull request description to have that text?

**Comment by ueshin** [comment]
@kayousterhout 
I'll update the unit test and JIRA, PR descriptions to use the simpler example.

As for the `createAncestorShuffleMapStages` way, I think it will have a risk of `StackOverflowError` for a long job because the master branch version of `getAncestorShuffleDependencies` finds descendants first, which would have a lot of ancestors with long linage, and if we immediately create the shuffle stage there by the `newOrUsedShuffleStage` method, which builds all ancestor stages by recursive-call fashion, the `StackOverflowError` will be thrown.

**Comment by kayousterhout** [comment]
I see -- now I understand the motivation for returning the shuffle dependencies topologically sorted, because it limits the depth of the recursion (it looks like the old code was trying to do that with the stack, but didn't quite get it right?).  Let me think about whether there's a simpler way to accomplish that.

**Comment by ueshin** [comment]
@kayousterhout 
I added a comment to [JIRA](https://issues.apache.org/jira/browse/SPARK-13902?focusedCommentId=15257721&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15257721).
Could you take a look at it and let me know which example I should use for the description, the original one or the "simpler" one.

**Comment by markhamstra** [comment]
It looks like https://github.com/apache/spark/pull/8923 by @suyanNone fixes at least the test that this PR adds to the DAGSchedulerSuite, and does so much more simply.

**Comment by kayousterhout** [comment]
Thanks for pointing that out @markhamstra. I'd be in favor of that solution (but augmented with a more clear test case -- which might just mean adding a nice ascii-art description of the DAG in the test case). What do you think Mark? @useshin do you see any issue with the simpler approach in #8923?

**Comment by markhamstra** [comment]
What I'm actually trying to look at right now is something very similar to https://github.com/apache/spark/pull/8923, but pushing into newOrUsedShuffleStage the check for prior existence of a Stage in shuffleToMapStage:
```scala
  private def newOrUsedShuffleStage(
                                      shuffleDep: ShuffleDependency[_, _, _],
                                      firstJobId: Int): ShuffleMapStage = {
    val rdd = shuffleDep.rdd
    val numTasks = rdd.partitions.length
    shuffleToMapStage.getOrElse(shuffleDep.shuffleId, {
      val stage = newShuffleMapStage(rdd, numTasks, shuffleDep, firstJobId, rdd.creationSite)
      if (mapOutputTracker.containsShuffle(shuffleDep.shuffleId)) {
        val serLocs = mapOutputTracker.getSerializedMapOutputStatuses(shuffleDep.shuffleId)
        val locs = MapOutputTracker.deserializeMapStatuses(serLocs)
        (0 until locs.length).foreach { i =>
          if (locs(i) ne null) {
            // locs(i) will be null if missing
            stage.addOutputLoc(i, locs(i))
          }
        }
      } else {
        // Kind of ugly: need to register RDDs with the cache and map output tracker here
        // since we can't do it in the RDD constructor because # of partitions is unknown
        logInfo("Registering RDD " + rdd.id + " (" + rdd.getCreationSite + ")")
        mapOutputTracker.registerShuffle(shuffleDep.shuffleId, rdd.partitions.length)
      }
      stage
    })
  }
```

**Comment by markhamstra** [comment]
I'd also like to address the concerns of @squito 's https://github.com/apache/spark/pull/8427 and @sitalkedia 's https://github.com/apache/spark/pull/12436

These PR's are all linked with the issue of needlessly duplicated Stages.  There probably needs to be some coordination in addressing the total set, but I hope that we can also handle them in more than one PR.

**Comment by kayousterhout** [comment]
@markhamstra why is the newOrUsedShuffleStage approach better / different?

IMO the existing code is very convoluted at the moment, but we should first fix the duplicate-stage bug (in the simplest way possible), and then in a separate PR address cleaning up some of the function calls so this whole thing is less convoluted and easier to reason about.  But I realize I may be missing some issue that the simple fix in #8923 doesn't address.

**Comment by markhamstra** [comment]
Yes, I'd also like to avoid trying to fix all of these related issues in one big PR.  There are overlapping concerns, but I agree that a simple fix to the duplicate-stage bug (SPARK-13902) is where I'd like to start -- which means delaying any merge of the other, related PRs [@rxin @andrewor14 @srowen @JoshRosen ... hopefully that's enough to get the message out].

I'd like to move the `shuffleToMapStage.getOrElse` within `newOrUsedShuffleStage` because that matches up better with the way I remember the desired semantics of newOrUsedStage -- i.e. if the desired Stage already exists, then re-use it; else create a new Stage.  Instead, what we do is: create a new Stage unconditionally; if some of the guts of the desired Stage already exist, then copy those over into the new Stage; else create a new Stage.  The way https://github.com/apache/spark/pull/8923 works is to retain the copy-the-guts approach  in newOrUsed, but to preface the callsite of newOrUsed with a guard to avoid the call in the Used case.  Seems like needlessly convoluted logic -- not that we're otherwise free from that particular sin within the DAGScheduler.

What I have now is kind of belt-and-suspenders. I don't know that the copy-the-guts logic within the `if (mapOutputTracker.containsShuffle(shuffleDep.shuffleId))` branch can ever do anything useful within the OrElse of `shuffleToMapStage.getOrElse`, so maybe that branch can be dropped.  I'm also still trying to convince myself that I can't lose my pants by not doing any mapOutput checking and copying in the case that `shuffleToMapStage` already has a reference to the desired used Stage.

**Comment by ueshin** [comment]
I saw the PR #8427 now.
Both the #8427 approach and @markhamstra's approach (should we use `getOrElseUpdate` instead of `getOrElse`?) seem like the simplest way to fix the duplicate-stage issue except a risk of `StackOverflowError`.

I agree that we should fix the duplicate-stage issue first in the simplest way possible, by #8427 or @markhamstra's, and then discuss a risk of `SOE` after one of them is merged.

**Comment by suyanNone** [comment]
as I know, duplicate stage occurs: 
stage 2 and stage 3  all  depends on stage 1
stage 4 depends on stage 2 and stage 3

So, if we get getAncestorShuffleDependencies(stage 4), it will depth-traverse stage 2-> stage 1, then traverse stage 3 -> stage 1... due to we use stack,  we can't de-duplicate the same stage in this.

if  we traverse stage 4 with de-duplication successful, it will in `shuffleIdToMapStage`, which can tell other rdd which stage had traversed. 
right?

**Comment by markhamstra** [comment]
I'm fine with the change in https://github.com/apache/spark/pull/8923  I'm looking at refactoring of `newOrUsedShuffleStage` that may effectively turn it into just `newShuffleStage`, so putting the `Used` check/guard in `getShuffleMapStage` where @suyanNone has it is a reasonable thing from my perspective.  Along with Kay's suggested augmentation of the tests, I think that's good enough for the first, simple PR.

**Comment by markhamstra** [comment]
Sorry, on re-reading my own post I realize that I wasn't completely clear.  I'm saying that the two-line change in DAGScheduler.scala from https://github.com/apache/spark/pull/8923 combined with Kay's requested test changes is all we need for a first PR.

**Comment by ueshin** [comment]
I think we are moving to #8923, but no updates there.
Should I use #8923 way here?

**Comment by kayousterhout** [comment]
@ueshin that would be great so we can merge this!

**Comment by ueshin** [comment]
@kayousterhout I simplified a test and updated JIRA and PR title and description.
Please take a look at them again and let me know what else to do to merge this PR.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 6]
```diff
@@ -322,6 +322,53 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with Timeou
...
+   *               <-------------
+   *
+   * Here, RDD B has a shuffle dependency on RDD A, and RDD C has shuffle dependency on both
+   * B and A. The shuffle dependency IDs are numbers in the DAGScheduler, but to make the example
+   * easier to understand, let's call the shuffled data from A shuffle dependency ID s_A and the
+   * shuffled data from B shuffle dependency ID s_B.
+   *
+   * Note: [] means an RDD, () means a shuffle dependency.
+   */
+  test("[SPARK-13902] not to create duplicate stage.") {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Can you change this to "[SPARK-13902] Ensure no duplicate stages are created"?

**Comment by kayousterhout** [comment]
This change looks good to me at this point (with the one small test name change). @markhamstra are you satisfied with this solution for now, and more significant clean up of this code path can be done in a later PR?

**Comment by markhamstra** [comment]
LGTM

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by andrewor14** [comment]
Sorry there was a build break in master.

retest this please

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
I took another look - we should probably merge this in 2.0 too. Kay can you do that? Thanks!

**Comment by kayousterhout** [comment]
Merged into 2.0.  Thanks @ueshin for your work on this and for bearing with us as we agreed on the simplest solution -- awesome to have this fixed!

**Comment by ueshin** [comment]
@kayousterhout, @markhamstra Thanks a lot!
I would like you to go back to #12060.
